
* hub
https://hgwdev.gi.ucsc.edu/~markd/kznf/deco-hub/hub.txt

* build:
cd data

# copy because they change
cp -pf uniprot/tab/*.9606.* . 

# GENCODE must get our own sequences, since transcripts rna fasta doesn't include alts
cp /hive/data/genomes/hg38/bed/gencodeV47/hgcImport/data/gencode.gp gencode.v47.gp
cp /hive/data/genomes/hg38/bed/gencodeV47/hgcImport/data/gencode.tsv gencode.v47.tsv

# filter for CDS
tawk '$6 < $7' gencode.v47.gp > gencode.v47.pc.gp
genePredToPsl /hive/data/genomes/hg38/chrom.sizes gencode.v47.pc.gp gencode.v47.pc.psl

# CDS must be upper-case
getRnaPred -cdsUpper -genomeSeqs=/hive/data/genomes/hg38/hg38.2bit no ./gencode.v47.pc.gp  all gencode.v47.pc.fa


* cross-species
Need chains from hg38 to target species.  Have them already for cls, so
just link
cd /hive/users/markd/gencode/projs/cls/cls-t2t-primate/hg38/data/assemblies

find . -name '*.chain.gz' -o -name '*.2bit' -o -name '*.sizes' | cpio -pduml /hive/users/markd/kznf/projs/bed-decorators/uniprotFeatMap/data/t2t-primates/

only use T2T apes, as need consistent CAT; drop musMus ratNor, calJac, macMul

rm -rf t2t-primates/musMus-GCA_000001635.* t2t-primates/ratNor-GCF_015227675.2/ t2t-primates/calJac-GCF_009663435.1/ t2t-primates/macMul-GCF_003339765.1/

* 
* notes:


* missing alignments

