<!DOCTYPE html>
<html lang="en">
  <head>
    <title>UniProt Transcript Annotations</title>
  </head>
  <body>
<h1>UniProt Transcript Annotations</h1>

<p>
  This <a href="https://genome.ucsc.edu/goldenPath/help/decorator.html"
          target="_blank">track decorator set</a> consist of UniProt annotations
  on individual GENCODE V45 protein-coding transcripts.
</p>
<p>
  Annotations from both the SwissProt and TrEMBL subsets of UniProt are
  available.  The UniProt canonical transcripts have annotations added to
  the transcript based on their protein coordinates. Non-UniProt canonical
  transcripts have the annotation coordinates projected via the genome coordinates.
</p>
<p>
  Disputed projected annotations are indicated in the display, as described below.
  This provides a visualization that a transcript may produce a protein that
  may have altered function or is non-functional.
</p>

<h2>Display Conventions and Configuration</h2>

<p>
  There are two types of annotations on transcripts, implemented as track
  decorators:
</p>
<ul>
  <li> Feature annotations: indicate the location of the UniProt features</li>
  <li> Disruption annotations: indicate the location of disruption in features projections</li>
</ul>

<h3>Feature annotations</h3>

<p>
  Feature annotations have their base items colored by type:
</p>
<ul>
  <li> <span style="color: rgb(255,0,150)"><b>signal peptide domains</b></span></li>
  <li> <span style="color: rgb(0,150,255)"><b>extracellular domains</b></span></li>
  <li> <span style="color: rgb(0,150,0)"> <b>transmembrane domains</b></span></li>
  <li> <span style="color: rgb(255,150,0)"><b>cytoplasmic domains</b></span></li>
  <li> <span style="color: rgb(100,0,0)"><b>topological domain</b></span></li>
  <li> <span style="color: rgb(100,100,0)"><b>zinc finger region</b></span></li>
  <li> <span style="color: rgb(100,100,100)"><b>disulfide bond</b></span></li>
  <li> <span style="color: rgb(12,12,120)"><b>lipoification modification sites</b></span></li>
  <li> <span style="color: rgb(0,100,100)"><b>glycosylation modification sites</b></span></li>
  <li> <span style="color: rgb(200,200,0)"><b>phosphorylation sites</b></span></li>
  <li> <span style="color: rgb(255,215,0)"><b>modified residues</b></span></li>
  <li> <span style="color: rgb(34,139,34)"><b>other SwissProt</b></span></li>
  <li> <span style="color: rgb(143,188,143)"><b>other TrEMBL</b></span></li>
</ul>

<p>
  Feature annotations are outlined based on whether they are UniProt canonical or non-canonical,
  or if the feature is disrupted:
</p>
<ul>
  <li> <span style="color: rgb(0,191,255)"><b>canonical</b></span>
  <li> <span style="color: rgb(0,0,0)"><b>noncanonical</b></span>
  <li> <span style="color: rgb(255,0,0)"><b>disrupted</b></span>
</ul>


<h3>Disruption annotations</h3>

<p>
Disruption annotations indicate where a feature is broken either by an
insertions or deletions create by alternative splicing. In addition to
outlining the disrupted features in <span style="color:rgb(255,0,0)">
<b>red</b></span>, additional annotations are added to show the location
in the feature where the disruption occurs.  These are represented by:
</p>
<ul>
  <li>A <span style="color:rgb(255,0,0)">
      <b>red</b></span> block annotation indicated the range where there is an insertion</li>
  <li>A <span style="color:rgb(255,0,0)">
      <b>red</b></span> triangle pointing to the points were a deletion
    occurred. Internal deletions will have two triangles pointing to each end of
    the deletion and truncations will have a single triangle.</li>
</ul>

<h3>Mouseover metadata</h3>

<ul>
  <li> UniProt data set: either <b>SwissProt</b> (manual annotation) or <b>TrEMBL</b> (automated annotation)
  <li> UniProt acc: UniProt accession with link to UniProt entry.
  <li> feature type: UniProt feature type
  <li> feature category: a more general category for the feature assigned by UCSC for filtering and display purposes
  <li> transcript category:
    <ul>
      <li><b>canonical</b>: transcript accession and version is listed by UniProt as match the protein
      <li><b>canonical_diff_version</b>: transcript accession is listed by UniProt as matching the protein however the
        version is different
      <li><b> non-canonical </b>: transcript that is not listed by UniProt that has the features projected on it
    </ul>
  <li> feature status:
    <ul>
      <li> for feature annotations:  <b>complete</b> or <b>disrupted</b>
      <li> for disruption annotations: <b>insert</b>, <b>del_5p</b> for
        5&apos; deletions, <b>del_3p</b> for 3&apos; deletions,
        or <b>del_int</b> for internal deletions
    </ul>
  <li> description: for feature annotations, description from UniProt; for disruption annotations,
    describes the type and size of the disruption.
  <li> comment: comment from UniProt
  <li> disease: disease name for disease-related variants
</ul>

<h3>Display control</h3>

<p>
  These controls allow the user to adjust the placement of the annotations.
  Block annotations (features and insertion disruptions) and glyph annotations
  (deletion disruptions) are controlled separately.
</p>
<ul>
  <li> <b>hide</b>: annotations are not shown
  <li> <b>overlay</b>: annotations are placed over the transcripts
  <li> <b>adjacent</b>: annotations are placed below the transcripts.
</ul>

<h3>Filters</h3>
<p>
  These filters are available to control what annotations are displayed.
  An annotation must match all filters to be displayed.
</p>
<ul>
  <li><b>Annotation category</b>: general category of annotation (default <b>Domain</b>)</li>
  <li><b>Uniprot dataset</b>: <b>SwissProt</b> or <b>TrEMBL</b> dataset (default <b>SwissProt</b>)</li>
  <li><b>Annotation type</b>: <b>feature</b> or <b>disruption</b> (default <b>All</b>)</li>
</ul>


<h2>Methods</h2>
<ol>
  <li> Convert the GENCODE transcript annotations on the genome to the implied alignments
    for use in projected (<em>transGenomeAln</em>)
    
  <li> Produce alignments of UniProt proteins to GENCODE mRNA transcripts
    (biotype XXX of protein_coding), resulting protein to mRNA alignments
    (<em>protTranAln</em>) with a two-step process:
    <ol>
      <li> UniProt proteins are aligned to their corresponding canonical mRNA
        transcript in GENCODE using translated BLAT. The UTR regions of the mRNA are
        hard-masked to prevent protein alignment to similar sequences outside of the CDS.  The
        canonical mRNA is matched, but the version number is ignored.
      <li> Produce protein-to-mRNA transcript alignments for the remaining
        protein-coding GENCODE transcripts that are not UniProt canonical
        transcripts by projecting the canonical protein to mRNA alignment
        through the genome.  Projections to UTR regions are clipped.
    </ol>
  <li> Project the UniProt feature annotations on the proteins to the associated mRNAs,
    and then project those to the genomic coordinates for each mRNA (<em>annotGenomeAln</em>).
    The projection pipeline is thus:
    <pre>
      annotation -&gt; protTranAln -&gt;  transGenomeAln =&gt; annotGenomeAln
    </pre>
  <li> Create a Track Decorator BED file associating the feature annotations
    map to each transcript with the transcript.  Associated metadata for each
    feature and identify disruptions in features.  Block decorations on the
    transcripts are created for each feature, and triangular glyph decorations
    are used to indicate the points of feature disruption.
</ol>

</html>

