<html>
  <head>
    <title>UniProt Transcript Annotations</title>
  </head>
  <body>
<h1>UniProt Transcript Annotations</h1>

<p>
  This <a href="https://genome.ucsc.edu/goldenPath/help/decorator.html"
          target="_blank">track decorator set</a> consist of UniProt annotations
  on individual GENCODE V45 protein-coding transcripts.
</p>
<p>
  Annotations from both the SwissProt and TrEMBL subsets of UniProt are
  available.  The UniProt canonical transcripts have annotations added to
  the transcript based on their protein coordinates. Non-UniProt canonical
  transcript have the annotation coordinates projected via the genome coordinates.
  Disputed projected annotations are indicated in the display, as described below.
</p>

<h2>Display Conventions and Configuration</h2>

There are two types of decorations for the transcripts: feature decorations,
that indicate where

<h4>Color coding:</h4>

Annotations have there base items colored by type:
<ul>
  <li> <span style="color: rgb(255,0,150)"><b>signal peptide domains</b></span>
  <li> <span style="color: rgb(0,150,255)"><b>extracellular domains</b></span>
  <li> <span style="color: rgb(0,150,0)"> <b>transmembrane domains</b></span>
  <li> <span style="color: rgb(255,150,0)"><b>cytoplasmic domains</b></span>
  <li> <span style="color: rgb(100,0,0)"><b>topological domain</b></span>
  <li> <span style="color: rgb(100,100,0)"><b>zinc finger region</b></span>
  <li> <span style="color: rgb(100,100,100)"><b>disulfide bond</b></span>
  <li> <span style="color: rgb(12,12,120)"><b>lipoification modification sites</b></span>
  <li> <span style="color: rgb(0,100,100)"><b>glycosylation modification sites</b></span>
  <li> <span style="color: rgb(200,200,0)"><b>phosphorylation sites</b></span>
  <li> <span style="color: rgb(255,215,0)"><b>modified residues</b></span>
  <li> <span style="color: rgb(34,139,34)"><b>other SwissProt</b></span>
  <li> <span style="color: rgb(143,188,143)"><b>other TrEMBL</b></span>
</ul>

Feature
<ul>
  <li> <span style="color: rgb()"><b></b></span>
  <li> <span style="color: rgb()"><b></b></span>
</ul>

    matched transcript outline           deepskyblue    0,191,255
    other isoform transcript outline     black          0,0,0

    feature insertion                    red            255,0,0
    feature deletion                     red            255,0,0
  <li> <span style="color: rgb()"><b></b></span>



-UUU:**-  F1  *uniprot*      Bot 

TrEMBL annotations are always shown in 
<span style="color: rgb(0,150,250)"><b>light blue</b></span>,
except in the Signal Peptides, Extracellular Domains, Transmembrane Domains,
and Cytoplamsic domains subtracks.</p>


    Mouse over a feature to see the full UniProt annotation comment. For variants, the mouse over will show the full name of the UniProt disease acronym.


<h2>Methods</h2>
<ol>
  <li> Convert the GENCODE transcript annotations on the genome to the implied alignments
    for use in projected (<em>transGenomeAln</em>)
    
  <li> Produce alignments of UniProt proteins to GENCODE mRNA transcripts
    (biotype XXX of protein_coding), resulting protein to mRNA alignments
    (<em>protTranAln</em>) with a two-step process:
    <ol>
      <li> UniProt proteins are aligned to their corresponding canonical mRNA
        transcript in GENCODE using translated BLAT. The UTR regions of the mRNA are
        hard-masked to prevent alignment of protein to similar sequences.  The
        canonical mRNA is matched ignoring version number.
      <li> Produce protein to mRNA transcript alignments for the remaining
        protein-coding GENCODE transcripts that are not UniProt canonical
        transcripts by projecting the canonical protein to mRNA alignment
        through the genome.  Projections to UTR regions are clipped.
    </ol>
  <li> Project the UniProt feature annotations on the proteins to the associated mRNAs,
    and then project those to the genomic coordinates for each mRNA (<em>annotGenomeAln</em>).
    The projection pipeline is thus:
    <pre>
      annotation -&gt; protTranAln -&gt;  transGenomeAln =&gt; annotGenomeAln
    </pre>
  <li> Create a Track Decorator BED file associating the feature annotations
    map to each transcript with the transcript.  Associate metadata for each
    feature and identify disruptions in features.  Block decorations on the
    transcripts are created for each feature, and triangular glyph decorations
    are used to indicate the points of feature disruption.
</ol>

</html>

