
* Alignment of protein to transcripts

Converting protein to transcript alignments to CDS to transcript alignment, by
multiplying the AA coordinates by three handles spliced codons once mapped to
genome.

Initially approach was to align proteins to both canonical and non-canonical
transcripts and pick the best alignments.  However, there were a number of
discrepancies in the non-canonical transcript alignments.  This was change to
only align to canonical and then project these alignments to the non-canonical
transcripts, which resulted in a significant improvement.

Alignments are done to canonical transcript where the UTR is hard-masked to
prevent alignment into UTR.  Projection to non-canonical transcripts have
UTR projection trimmed.

Both BLAT protein to DNA and TBLASTN were tried to align the UniProt
proteins.  When aligning only to the canonical transcript, the difference
were negligible. Could probably avoid a lot of the BLAT alignments by
check if the transcript version was the same and just construct it.

An optimization would be to avoid alignments to the canonical mRNA
when there is an exact match.

* Cross-species projection of UniProt/SwissProt annotations
Native TrEMBL annotations at available for many species and these can be
mapped to the genome using this pipeline. SwissProt annotations, when
available, are more extensive and can be mapped using cross-species genomic
alignments

* InterProScan
InterProScan is run directly on proteins generated from the annotations, thus
an alignment step is unnecessary.

* Cross-species comparison
Projection (TransMap) alignments of SwissProt from a reference genome to the
target genome can find deleted and modified domains; however the do not
identify novel domains.  Comparison with InterProScan may fine novel insertions.

** Comparison is done by:
 - Associate each target mRNAs annotation with the TransMap alignment of the
   source genome mRNAs with the closest similarity.  A given source annotation
 - Compare domain annotation with source and target annotations, reporting
   changes
 - Mapped SwissProt annotations are more comprehensive than native TrEMBL or
   InterProScan
   
** Details:
*** src annotation to src mRNA alignments
- input from uniprotAnnotsMap
    - annotTrans.swissprot.src.psl
    - annotGenome.swissprot.mapped.psl
*** target annotation to mRNA alignments
- input from uniprotAnnotsMap native (TrEMBL) and InterProScan
    - annotTrans.target.psl
    - annotGenome.target.psl

*** Associate each target mRNA with the most similar src mRNA

*** scan through paired mRNAs and note differences in annotations
**** mapping via implicit RNA/RNA aligns in target genome
srcAnnot-srcTrans -> 
srcTrans-targetGenome <-> targetGenome-targetTrans
<- targetTrans-targetAnnot
***** implicit means mapping to the same target genome
***** src annotation files
uniprotMetaTsv
annot2GenomePsl
annot2TransRefTsv
xspeciesTrans2TransPsl
***** target annotation files

**** Collect differences based on similarity
Collection is done based on annotations pairing
defined in a configuration file.
For SwissProt mapped, this is the same annotation type,
for InterProScan, association is needed, which maybe
one to many.
