* read-through genes not handled well

* is "canonical" a good term for the matched GENCODE mRNA

* does canonical to canonical mRNA self-alignment ever produce something
that isn't 100%

* look at miniprot

* doUniprot make featType more readable,
maybe add this to comment?
if annot.featType == "sequence variant":
    annoType = "Naturally occurring sequence variant"
elif annot.featType == "mutagenesis site":
    annoType = "Experimental mutation of amino acids"

* uniprotIsoCanonicalAlign
Has this note:
- Don't filter for minimum alignment coverage to get alt isoforms, which are often split
but doe filter some.  Maybe these alignments should be chained

* should uniprotIsoCanonicalAlign and uniprotIsoCanonicalSelect be combined?
* should proteinTranscriptAlign and uniprotGencodeSelect be combined

* TODO add source transcript
* TODO how to flag deleted domains

* TFDP2: no annotations
* ENST00000235310.7 (in tests) seems to get multiple similar domains

* ZNF649 broken blast and blat alignments

* ZNF429 chr19:21,499,700-21,559,736
ENST00000597078.5 blast missing KRAB

* ZNF534 chr19:52,429,148-52,442,499
blast misses incomplete KRABS found by blat


* intersect with Max tracks to find issues
