#+STARTUP: nologdone
#+TODO: TODO ACTIVE IMPORTANT URGENT PAUSED | DONE CANCELED

* Make smaller triangle* color other red
* color-blind safe

* colors:
** other than special color cases, can TrEMBL uses a different shade or outline

* genomic uniprot color issues
** lipoification sites
is 12,12,120, which is default swissprot color, but not explicitly set
** these are not doced in genomic page
*** modified residue                     goldenrod      200,200,0
*** disulfide bond                       dimgray        100,100,100
*** topological domain                   maroon         100,0,0
*** zinc finger region                   olive          100,100,0

** phosphorylation site color 200,200,0 is used for other types of modified residues
should look for  (phosphoserine, phosphothreonine and phosphotyrosine) in comment

 <li> <span style="color: rgb(200,200,0)"><b>phosphorylation sites</b></span>
  <li> <span style="color: rgb(200,200,0)"><b>modified residues</b></span>

 In the "UniProt Modifications" track, lipoification sites are highlighted in dark blue, glycosylation sites in dark
green, and phosphorylation in light green. 

Common modifications include phosphorylation, methylation, acetylation, amidation, formation of pyrrolidone carboxylic acid, isomerization, hydroxylation, sulfation, flavin-binding, cysteine oxidation and nitrosylation.


* remove "phosphorylation site" FIXME code

* Should indicate if different version of canonical used

* big insertion rectangle is weird
chr1:11,674,479-11,681,148
due to one exon being different

* python errors in ../bin/uniprotAnnotsToDecorators
cause it to continue to run lots of threads, creating a huge stack trace

* URGENT why didn't split alignment block get chained
in test set at chr19:51,890,708-51,890,879 ENST00000600738.5

** gencode protein alignment
uniprotProteinTranscriptAlign - makes aligns, maybe change to only align matched transcripts
uniprotGencodeSelect - filters alignments, maybe change to build non-matched by projection
or should these be merged


* URGENT need to check multi-cystronic gene handling
CDKN2A: P42771 and Q8N726

* URGENT Other uniprot domain types
** should these be classified as domains or filtered by default:
https://www.uniprot.org/keywords/KW-9994

** P42771 - repeat ANK 1..4
https://www.uniprot.org/keywords/KW-0040



* It was less than satisfying using Pandas, so it was removed with a huge speedup
with pandas, time of test uniprotAnnotsToDecorators --nprocs=1
real	0m6.944s
user	0m8.726s
sys	0m3.686s

without pandas
real	0m0.918s
user	0m0.804s
sys	0m0.055s


* read-through genes not handled well
could be handled by projection annotations

* does canonical to canonical mRNA self-alignment ever produce something
that isn't 100%

* look at miniprot

* doUniprot make featType more readable,
maybe add this to comment?
if annot.featType == "sequence variant":
    annoType = "Naturally occurring sequence variant"
elif annot.featType == "mutagenesis site":
    annoType = "Experimental mutation of amino acids"


* percent-identity doesn't seem very sane from blastToPsl for protein to RNH gene (RNH1?)

* should proteinTranscriptAlign and uniprotGencodeSelect be combined

* TODO add source transcript
* TODO how to flag deleted domains

* TFDP2: no annotations
* ENST00000235310.7 (in tests) seems to get multiple similar domains

* ZNF649 broken blast and blat alignments

* ZNF429 chr19:21,499,700-21,559,736
ENST00000597078.5 blast missing KRAB

* ZNF534 chr19:52,429,148-52,442,499
blast misses incomplete KRABS found by blat


* intersect with Max tracks to find issues

* convert from prMsg to using python loggin

* in uniprotMapAnnots, it seems it is should check for end past end protein:

Warning: Q9BXI3 annotation extends past end of protein: 369 > 368
Warning: Q96P26-1 annotation extends past end of protein: 611 > 610
Warning: Q96P26-1 annotation extends past end of protein: 611 > 610
Warning: Q9H0P0-4 annotation extends past end of protein: 337 > 336
 
* URGENT uniprotMapAnnots warnings should maybe be written to problem file; also why is does not map repeat?
Warning: P30511-1 annotation extends past end of protein: 353..354 > 346
Warning: Q99732-1 annotation extends past end of protein: 174..175 > 161
Warning: P20848 does not map to any transcript
Warning: P20848 does not map to any transcript
Warning: P20848 does not map to any transcript
Warning: P20848 does not map to any transcript
Warning: P20848 does not map to any transcript
Warning: P20848 does not map to any transcript
Warning: P20848 does not map to any transcript

  
* default to adjacent

* Interesting cases
** chr1:11,658,702-11,658,804
frame-shifted protein; looks good, but this the display ideal?

* should other classes besides domains be defaulted
** BRAC2
P51587 - SwissProt
no annotations classified as "domain", but have repeat

* should be no need to pre-filter GENCODE metadata, annotations and alignments for protein coding:
they should just be ignored
see bigtest/bigtest.org

* IMPORTANT Adam F. feedback: domain ends might a bit unsure, so indicate the amount of truncation.
use orange or yellow?

* better color for other iosform okay status?
black looks weird

* look a problem report cases

* option to exclude CDS truncated cases
useful in visual QC to look for weird cases

* mark both ends of internal breakage

* issues:
** chr19:51,877,172-51,877,490 ENST00000451628.9
why is domain truncated here
also says 5' truncation, looks like 3'

