* read-through genes not handled well

* doUniprot make featType more readable,
maybe add this to comment?
if annot.featType == "sequence variant":
    annoType = "Naturally occurring sequence variant"
elif annot.featType == "mutagenesis site":
    annoType = "Experimental mutation of amino acids"

* uniprotIsoCanonicalAlign
Has this note:
- Don't filter for minimum alignment coverage to get alt isoforms, which are often split
but doe filter some.  Maybe these alignments should be chained

* could uniprotIsoCanonicalAlign and uniprotIsoCanonicalSelect be combined?
