#+STARTUP: nologdone
#+TODO: TODO ACTIVE URGENT PAUSED | DONE CANCELED

* URGENT why didn't split alignment block get chained
in test set at chr19:51,890,708-51,890,879 ENST00000600738.5

** gencode protein alignment
uniprotProteinTranscriptAlign - makes aligns, maybe change to only align matched transcripts
uniprotGencodeSelect - filters alignments, maybe change to build non-matched by projection
or should these be merged


* URGENT need to check multi-cystronic gene handling
CDKN2A: P42771 and Q8N726

* URGENT Other uniprot domain types
** should these be classified as domains or filtered by default:
https://www.uniprot.org/keywords/KW-9994

** P42771 - repeat ANK 1..4
https://www.uniprot.org/keywords/KW-0040



* It was less than satisfying using Pandas, so it was removed with a huge speedup
with pandas, time of test uniprotAnnotsToDecorators --nprocs=1
real	0m6.944s
user	0m8.726s = 1206s
sys	0m3.686s

without pandas
real	0m0.918s
user	0m0.804s
sys	0m0.055s


* read-through genes not handled well
could be handled by projection annotations

* is "canonical" a good term for the matched GENCODE mRNA

* does canonical to canonical mRNA self-alignment ever produce something
that isn't 100%

* look at miniprot

* doUniprot make featType more readable,
maybe add this to comment?
if annot.featType == "sequence variant":
    annoType = "Naturally occurring sequence variant"
elif annot.featType == "mutagenesis site":
    annoType = "Experimental mutation of amino acids"


* percent-identity doesn't seem very sane from blastToPsl for protein to RNH gene (RNH1?)

* should proteinTranscriptAlign and uniprotGencodeSelect be combined

* TODO add source transcript
* TODO how to flag deleted domains

* TFDP2: no annotations
* ENST00000235310.7 (in tests) seems to get multiple similar domains

* ZNF649 broken blast and blat alignments

* ZNF429 chr19:21,499,700-21,559,736
ENST00000597078.5 blast missing KRAB

* ZNF534 chr19:52,429,148-52,442,499
blast misses incomplete KRABS found by blat


* intersect with Max tracks to find issues

* convert from prMsg to using python loggin

* in uniprotMapAnnots, it seems it is should check for end past end protein:

Warning: Q9BXI3 annotation extends past end of protein: 369 > 368
Warning: Q96P26-1 annotation extends past end of protein: 611 > 610
Warning: Q96P26-1 annotation extends past end of protein: 611 > 610
Warning: Q9H0P0-4 annotation extends past end of protein: 337 > 336
 
* URGENT uniprotMapAnnots warnings should maybe be written to problem file; also why is does not map repeat?
Warning: P30511-1 annotation extends past end of protein: 353..354 > 346
Warning: Q99732-1 annotation extends past end of protein: 174..175 > 161
Warning: P20848 does not map to any transcript
Warning: P20848 does not map to any transcript
Warning: P20848 does not map to any transcript
Warning: P20848 does not map to any transcript
Warning: P20848 does not map to any transcript
Warning: P20848 does not map to any transcript
Warning: P20848 does not map to any transcript

  
* default to adjacent

* rather than special color cases, can TrEMBL uses a different shade or outline

* Interesting cases
** chr1:11,658,702-11,658,804
frame-shifted protein; looks good, but this the display ideal?

* no swissprot domains mappings
** BRAC2
P51587 - SwissProt
no annotations classified as "domain", but have repeat

* should be no need to pre-filter GENCODE metadata, annotations and alignments for protein coding:
they should just be ignored
see bigtest/bigtest.org

* modify uniprotProteinTranscriptAlign to only blat/blast to matched transcripts
