#+STARTUP: nologdone
#+TODO: TODO URGENT | DONE CANCELED

* uniprot.py  diff between id, acc, main isoform acc|id not clear

* annotation term used when feature is UniProt terminology

* AnnotTransRef
** is transPostion really needed, given we have the psl
** canonTransId not implement, but is it needed

* Max:
** In "adjavent" mode, I'm very tempted to click on features... we could have
an hgc page for these annotations, it's a bigBed file after all...


** DONE I don't understand this:
> feature status: disrupted
Is this UniProt data? We usually only show what Uniprot provides, but I
imagine this was added by your pipeline? Wonder if we should rename this or
somehow highlight that this is calculated information?

** The triangles are too big. I guess we can just shed off a few pixels. Maybe
depending on zoom levels?

** The triangles are not intuitive to me at all. A little message in the
mouseover may help?

** DONE OK a tiny one: The mouseovers start with:
> UniProt data set: SwissProt
> UniProt acc: P00395
But to users, that's not super relevant... I'd have the more human-readable
fields shown first.


* Make smaller triangle* color other red
** color-blind safe

* colors:
** other than special color cases, can TrEMBL uses a different shade or outline
** is UniProtCategory still needed if TrEMBL coloring changed

* genomic uniprot color issues
** lipoification sites
is 12,12,120, which is default swissprot color, but not explicitly set

** phosphorylation site color 200,200,0 is used for other types of modified residues
should look for  (phosphoserine, phosphothreonine and phosphotyrosine) in comment

 <li> <span style="color: rgb(200,200,0)"><b>phosphorylation sites</b></span>
  <li> <span style="color: rgb(200,200,0)"><b>modified residues</b></span>

 In the "UniProt Modifications" track, lipoification sites are highlighted in dark blue, glycosylation sites in dark
green, and phosphorylation in light green. 

Common modifications include phosphorylation, methylation, acetylation, amidation, formation of pyrrolidone carboxylic acid, isomerization, hydroxylation, sulfation, flavin-binding, cysteine oxidation and nitrosylation.




* big insertion rectangle is weird
chr1:11,674,479-11,681,148
due to one exon being different

* python errors in ../bin/uniprotAnnotsToDecorators
cause it to continue to run lots of threads, creating a huge stack trace

* URGENT why didn't split alignment block get chained
in test set at chr19:51,890,708-51,890,879 ENST00000600738.5


* URGENT need to check multi-cystronic gene handling
CDKN2A: P42771 and Q8N726

* URGENT Other uniprot domain types
** should these be classified as domains or filtered by default:
https://www.uniprot.org/keywords/KW-9994

** P42771 - repeat ANK 1..4
https://www.uniprot.org/keywords/KW-0040



* It was less than satisfying using Pandas, so it was removed with a huge speedup
with pandas, time of test uniprotAnnotsToDecorators --nprocs=1
real	0m6.944s
user	0m8.726s
sys	0m3.686s

without pandas
real	0m0.918s
user	0m0.804s
sys	0m0.055s


* read-through genes not handled well
could be handled by projection annotations

* does canonical to canonical mRNA self-alignment ever produce something
that isn't 100%

* look at miniprot

* doUniprot make featType more readable,
maybe add this to comment?
if annot.featType == "sequence variant":
    annoType = "Naturally occurring sequence variant"
elif annot.featType == "mutagenesis site":
    annoType = "Experimental mutation of amino acids"


* should proteinTranscriptAlign and uniprotGencodeSelect be combined

* TODO how to flag deleted domains

* TFDP2: no annotations
* ENST00000235310.7 (in tests) seems to get multiple similar domains
looks okay, but interesting test acse

* intersect with Max tracks to find issues

* convert from prMsg to using python logging

  
* default to adjacent

* Interesting cases
** chr1:11,658,702-11,658,804
frame-shifted protein; looks good, but this the display ideal?

* should other classes besides domains be defaulted
** BRAC2
P51587 - SwissProt
no annotations classified as "domain", but have repeat

* should be no need to pre-filter GENCODE metadata, annotations and alignments for protein coding:
they should just be ignored
see bigtest/bigtest.org

* IMPORTANT Adam F. feedback: domain ends might a bit unsure, so indicate the amount of truncation.
use orange or yellow?

* better color for other iosform okay status?
black looks weird

* look a problem report cases

* option to exclude CDS truncated cases
useful in visual QC to look for weird cases

* mark both ends of internal breakage

* issues:
** chr19:51,877,172-51,877,490 ENST00000451628.9
why is domain truncated here
also says 5' truncation, looks like 3'

* NOTCH2NLB has annotation disrupted in canonical
could be because it is wrong because of bad annotation; document
