
* /hive/data/outside/otto/uniprot/
doUniprot
makeUniProtPsl.sh

* hub
 https://hgwdev.gi.ucsc.edu/~markd/kznf/deco-hub/hub.txt


 USER=demo-markd jkmmake cgi >&log
 https://hgwdev-demo-markd.gi.ucsc.edu/
* 

* gencode:

* data

* samples

* geneName

* algorithm:
- gencode protein coding RNAs
- align sp/tr to RNAs
- for sp:
  for each transcript:
  - if there is a SP that is associated with a transcript:
    use the best alignment to transcript
  - if there is not an SP associated with transcript, then select
    the best alignment of another SP associated with the same gene.
  
  group alignments and isoforms by gene:
  
  
  filter based on either
  hgncSym or hgncId,  ensemblGene, ensemblTrans

* including refseq
/hive/data/genomes/hg38/bed/ncbiRefSeq.2024-01-29/process/


* metadata pass through
* design of pass passing through unmapped features:
properly order between other features
can not be a PSL as there is nothing aligned

*** load UniProt tab data into sqlite data, much easier to make parallel

** uniprotAnnotsMap
*** thinking:
- rather than annotation using PSL line, unmapped features still get ids
- is it enough to anchor ids or do need to create 1-bases psl with relative locations

- unmapped PSL are passed throuth, dropped ones added back
- updated metadata produced

** metadata protein-transcript map
uniprotId  status

"no_gene_transcript_in_geneset"
"no_transcript_in_geneset"
"uniprot_not_aligned"
"noncanon_trans_not_mapped"

*** new
uniprotAcc
status
taxonId
dataSet
transcriptId
transCategory
canonTransId
cdsCoverage



** metadata annotMap
annotId	annotMapId	transcriptPos	transcriptId	canonTransId	alignIdx

canonTransId not filled in  
